### ISSUE TITLE
Certificate Rotation for Landscaper components

### ISSUE THREAD

**Issue by danielfoehrKn** [gardener/gardener#4856]
**How to categorize this issue?**

/area security
/kind enhancement
/priority 2

**What would you like to be added**:

We need a strategy to automatically rotate Certificates used during a Gardener landscape deployment.
This was done manually in the past.

I think we need to distinguish between two categories
1) externally provided certificate (e.g. supplied as a secret reference and rotated by an external Cert-Manager)
2) certificated generated by the landscaper components themselves

This issue is only concerned with case 2). 

However, it is not necessarily clear in what category a certificate falls.

**Example**:  Control Plane component
Imports: GardenerCA certificate
 - as direct import values
 - as secret reference
 - if missing: imports from existing installation

Exports: GardenerCA certificate

Let's say on the initial run we do not supply a Gardener CA certificate so that the landscaper component generates one. The newly generated CA is exported. The CA is backed up to git /s3.
On a subsequent execution, the previously exported CA is supplied as import values again
- why: to prevent a CA re-generation only because someone deletes the secret in the APIService in the installation 
How can the component now distinguish if the CA certificate is managed externally or has been generated by itself before?

**Possible solution**:  Each imported certificate has a flag `managed` that indicates whether it is managed by the landscaper component or not. Only `managed` certificates are rotated.

The component has the following global configuration
```
type CertificateRotation struct {
	// Enabled defines if certificate rotation is enabled
	Enabled bool
	// ValidityThresholdPercentage defines the threshold when a certificate shall be rotated
	// Defaults to 80%
	// Example: rotates the CA certificate and it's depending TLS certificates if >= 80% of the CA certs lifetime is reached
	ValidityThresholdPercentage float64
}
```

In order to rotate
 - rotation flags for cert must be true
 - validity threshold is expired

The landscaper component then re-generates a CA and its dependent TLS serving certificates, deploys them into the Cluster, and writes to the export file.
 The user must make sure that the old certificate is not imported on the next execution. Otherwise: regeneration loop
 - import newly generated certificate previously exported in the next run! 
 - do not provide the newly generated cert - component will retrieve it from the cluster. Danger: what happens if deleted by accident from the cluster? -> will re-generate.
- this [could be done by a generic backup solution in the landscaper.](https://github.com/gardener/landscaper/issues/307)

Prerequisites 
- landscaper components are no controllers. Hence, it is vital for them to be executed regularly so that they can recognize that a cert is about to expire. The landscaper should automatically ensure reconciliation of Installations at least once a day (similar to maintenance time in a Shoot cluster).

Related point:
 - configurable lifetime of Generated TLS and CA certificates. This should be done after automatic rotation is possible.

**Alternative 1**:

Each landscaper component has a `rotateNow` flag that re-generates all certificates. 
 - that are not provided in its import configuration
 - that have flag `managed` == true

Advantage:
 - simplicity

Downside: Manual effort. 
- Need to set `rotateNow` to true in Installation's import config and trigger reconciliation of installation.
- Manually keep track of expiring certificates (not good if we want to encourage low certificate lifetimes)

**Alternative 2**:

Certificates are all externally managed e.g by the landscaper or another component.
While components would generate certificates initially if not provided in their import configuration, an external controller would keep track of certificates to rotate.
E.g by an annotation on secrets

Advantage:
 - Not each component has to deal with it separately.

Downsides:
 - how does that work for certs that are e.g stored in an APIService (CA)? The external component must be able to support all these resources.
 - How to update a certificate stored in secret X in Cluster A that needs to be written to e.g `ValidatingWebhookConfiguration` in Cluster B (example: gardenloginController cc @petersutter ). Similar to what jetstack cert-manager does with injection via annotation.
 - External component must be able to generate CA certificates (can jetstack cert-manager do that?)

**Why is this needed**:

Manual certificate rotation is hard and is therefore mostly avoided by creating certificates with dangerously long lifetimes (e,g years or decades).

**Comment by danielfoehrKn** [comment]
cc @petersutter @achimweigel @robertgraeff

**Comment by danielfoehrKn** [comment]
> Alternative 2: Certificates are all externally managed e.g by the landscaper or another component.

It has been decided that landscaper components should rotate certificates themselves. This is because it is hard to build a central service that knows how to rotate certificates and store them in the appropriate location (e.g APIService) for each component.

> Alternative 1: Each landscaper component has a rotateNow flag that re-generates all certificates.

At least the control plane component is built to include such a flag. However, it would be probably nice to have a flag for all components so that the user does not have to set the `rotation` flag for each component.

**Comment by danielfoehrKn** [comment]
> The landscaper component then re-generates a CA and its dependent TLS serving certificates, deploys them into the Cluster, and writes to the export file.
The user must make sure that the old certificate is not imported on the next execution. Otherwise: regeneration loop

The most straightforward solution currently seems to me to make it the "users" responsibility to supply the rotated certificates in the import configuration for the next execution. 
Each landscaper component should specify in its export configuration whether a certificate has been rotated or not.

However, in general, landscaper components should be built to synchronize against an existing installation. This means existing certificates stored in a Gardener installation (e.g in secret in the garden cluster) should be re-used instead of regenerated.
This would allow completely hands-off certificate rotations as, during the next execution of the landscaper component, the rotated secret is again synchronized from the existing Gardener installation.

The only exception is when there is no existing Gardener installation yet and custom certificates shall be used. Then after the first execution, the custom-provided certificates should be removed from the import configuration (to avoid overwriting rotated certificates already stored in cluster)


We should guarantee that Gardener certificates are backed up (e.g the CA of the Gardener API server).
Instead of backing up the certificates in the export configuration of each component [using a landscaper-specific solution](https://github.com/gardener/landscaper/issues/307), I would prefer a generic solution.
When synchronizing against an existing installation, we store all certificates in the Kubernetes cluster itself (instead of e.g git). We can use [Velero](https://github.com/vmware-tanzu/velero) to backup the secrets of an existing Gardener installation. 

I am currently building the control-plane component to demonstrate certificate re-use from an existing installation to allow "hands off" certificate rotations. 

@enrico-kaack-comp  to make you aware of the certificate rotation topic for landscaper components.

**Comment by danielfoehrKn** [comment]
This issue contains brainstorming regarding certificate rotation for landscaper components. 
There does not seem to be any traction/discussion happening.
So for now, I regard the control plane component as the "reference component" of how certificate rotation should be implemented for landscaper components.

Hence, this issue can be closed once the linked PR is merged.

Further, I would use an etcd backup for the runtime cluster etcd + virtual garden etcd. 
But this is independent of the cert rotation topic and is not relevant for OSS Gardener.
